src/porepy/viz/exporter.py

```diff
@@ -520,546 +1064,789 @@ def _export_grid(
+
+                # Find all cells with n faces, and store for later use
+                cells = np.nonzero(num_faces_per_cell == n)[0]
+                g_cell_map[cell_type] = cells
+
+                # Store cell ids in global container; init if entry not yet established
+                if cell_type not in cell_id:
+                    cell_id[cell_type] = []
+
+                # Add offset taking into account previous grids
+                cell_id[cell_type] += (cells + cell_offset).tolist()
+
+            # Determine cell-node connectivity for each cell type and all cells.
+            # Treat triangle, quad and polygonal cells differently
+            # aiming for optimized performance.
+            for n in np.unique(num_faces_per_cell):
+
+                # Define the cell type
+                cell_type = polygon_map.get(f"polygon{n}", f"polygon{n}")
+
+                # Check if cell_type already defined in cell_to_nodes, otherwise construct
                 if cell_type not in cell_to_nodes:
-                    cell_to_nodes[cell_type] = np.atleast_2d(nodes_loc[0] + pts_pos)
-                    cell_id[cell_type] = [cell_pos]
-                else:
-                    cell_to_nodes[cell_type] = np.vstack(
-                        (cell_to_nodes[cell_type], nodes_loc[0] + pts_pos)
-                    )
+                    cell_to_nodes[cell_type] = np.empty((0, n), dtype=int)
+
+                # Special case: Triangle cells, i.e., n=3.
+                if cell_type == "triangle":
 
-                    cell_id[cell_type] += [cell_pos]
-                cell_pos += 1
+                    # Triangles are simplices and have a trivial connectivity.
+
+                    # Fetch triangle cells
+                    cells = g_cell_map[cell_type]
+                    # Determine the trivial connectivity.
+                    cn_indices = self._simplex_cell_to_nodes(3, g, cells)
+
+                # Quad and polygon cells.
+                else:
+                    # For quads/polygons, g.cell_nodes cannot be blindly used as for
+                    # triangles, since the ordering of the nodes may define a cell of
+                    # the type (here specific for quads) x--x and not x--x
+                    #  \/          |  |
+                    #  /\          |  |
+                    # x--x         x--x .
+                    # Therefore, use both g.cell_faces and g.face_nodes to make use of face
+                    # information and sort those to retrieve the correct connectivity.
+
+                    # Strategy: Collect all cell nodes including their connectivity in a
+                    # matrix of double num cell size. The goal will be to gather starting
+                    # and end points for all faces of each cell, sort those faces, such that
+                    # they form a circlular graph, and then choose the resulting starting
+                    # points of all faces to define the connectivity.
```

Comment:
Excellent documentation! Should we consider extracting a subfunction?

Reply:
No need for a subfunction until we need the ordering somewhere else. I hope that will not be the case for a while.